consensus/qbft/core/core.go

```diff
@@ -0,0 +1,304 @@
+	c := &core{
+		config:             config,
+		address:            backend.Address(),
+		state:              StateAcceptRequest,
+		handlerWg:          new(sync.WaitGroup),
+		logger:             log.New("address", backend.Address()),
+		backend:            backend,
+		backlogs:           make(map[common.Address]*prque.Prque),
+		backlogsMu:         new(sync.Mutex),
+		pendingRequests:    prque.New(),
+		pendingRequestsMu:  new(sync.Mutex),
+		consensusTimestamp: time.Time{},
+		roundMeter:         metrics.NewMeter(),
+		sequenceMeter:      metrics.NewMeter(),
+		consensusTimer:     metrics.NewTimer(),
+	}
+
+	// During upgrades we do not want to re-register round, sequence and consensus metrics
+	// Do not have to register them, if they are already registered
+	if roundMeter := r.Get("consensus/istanbul/core/round"); roundMeter == nil {
+		r.Register("consensus/istanbul/core/round", c.roundMeter)
+	}
+	if sequenceMeter := r.Get("consensus/istanbul/core/sequence"); sequenceMeter == nil {
+		r.Register("consensus/istanbul/core/sequence", c.sequenceMeter)
+	}
+
+	if consensusTimer := r.Get("consensus/istanbul/core/consensus"); consensusTimer == nil {
+		r.Register("consensus/istanbul/core/consensus", c.consensusTimer)
+	}
+
+	c.validateFn = c.checkValidatorSignature
+	return c
+}
+
+// ----------------------------------------------------------------------------
+
+type core struct {
+	config  *istanbul.Config
+	address common.Address
+	state   State
+	logger  log.Logger
+
+	backend               istanbul.Backend
+	events                *event.TypeMuxSubscription
+	finalCommittedSub     *event.TypeMuxSubscription
+	timeoutSub            *event.TypeMuxSubscription
+	futurePreprepareTimer *time.Timer
+
+	valSet     istanbul.ValidatorSet
+	validateFn func([]byte, []byte) (common.Address, error)
+
+	backlogs   map[common.Address]*prque.Prque
+	backlogsMu *sync.Mutex
+
+	current   *roundState
+	handlerWg *sync.WaitGroup
+
+	roundChangeSet   *roundChangeSet
+	roundChangeTimer *time.Timer
+
+	QBFTPreparedPrepares []*message.SignedPreparePayload
+
+	pendingRequests   *prque.Prque
+	pendingRequestsMu *sync.Mutex
+
+	consensusTimestamp time.Time
+	// the meter to record the round change rate
+	roundMeter metrics.Meter
+	// the meter to record the sequence update rate
+	sequenceMeter metrics.Meter
+	// the timer to record consensus duration (from accepting a preprepare to final committed stage)
+	consensusTimer metrics.Timer
+}
+
+func (c *core) currentView() *istanbul2.View {
+	return &istanbul2.View{
+		Sequence: new(big.Int).Set(c.current.Sequence()),
+		Round:    new(big.Int).Set(c.current.Round()),
+	}
+}
+
+func (c *core) IsProposer() bool {
+	v := c.valSet
+	if v == nil {
+		return false
+	}
+	return v.IsProposer(c.backend.Address())
+}
+
+func (c *core) IsCurrentProposal(blockHash common.Hash) bool {
+	return c.current != nil && c.current.pendingRequest != nil && c.current.pendingRequest.Proposal.Hash() == blockHash
+}
+
+func (c *core) commitQBFT() {
```

Comment:
Should `commitQBFT` be moved to `commit.go` ?

Reply:
Done